src/Fixer/Import/FullyQualifiedStrictTypesFixer.php

```diff
@@ -447,17 +447,40 @@ private function setupUsesFromDiscoveredSymbols(array &$uses, string $namespaceN
                 $discoveredFqcnByShortNameLower[strtolower($useShortName)] = $useLongName;
             }
 
-            uasort($discoveredSymbols, static fn ($a, $b) => substr_count($a, '\\') <=> substr_count($b, '\\'));
+            $useByShortNameLower = [];
+            foreach ($uses as $useShortName) {
+                $useByShortNameLower[strtolower($useShortName)] = true;
+            }
+
+            uasort($discoveredSymbols, static function ($a, $b) {
+                $res = str_starts_with($a, '\\') <=> str_starts_with($b, '\\');
+                if (0 !== $res) {
+                    return $res;
+                }
+
+                return substr_count($a, '\\') <=> substr_count($b, '\\');
+            });
             foreach ($discoveredSymbols as $symbol) {
-                $shortEndNameLower = strtolower(str_contains($symbol, '\\') ? substr($symbol, strrpos($symbol, '\\') + 1) : $symbol);
-                if (!isset($discoveredFqcnByShortNameLower[$shortEndNameLower])) {
-                    if ('' !== $namespaceName && !str_starts_with($symbol, '\\') && str_contains($symbol, '\\')) { // @TODO add option to force all classes to be imported
-                        continue;
+                while (true) {
+                    $shortEndNameLower = strtolower(str_contains($symbol, '\\') ? substr($symbol, strrpos($symbol, '\\') + 1) : $symbol);
+                    if (!isset($discoveredFqcnByShortNameLower[$shortEndNameLower])) {
+                        $shortStartNameLower = strtolower(explode('\\', ltrim($symbol, '\\'), 2)[0]);
+                        if (str_starts_with($symbol, '\\') || ('' === $namespaceName && !isset($useByShortNameLower[$shortStartNameLower]))
+                            || !str_contains($symbol, '\\')
+                        ) {
+                            $discoveredFqcnByShortNameLower[$shortEndNameLower] = $this->resolveSymbol($symbol, $uses, $namespaceName);
+
+                            break;
+                        }
+                    }
+                    // else short name collision - keep unimported
+
+                    if ((str_starts_with($symbol, '\\') || '' === $namespaceName) || !str_contains($symbol, '\\')) {
+                        break;
                     }
 
-                    $discoveredFqcnByShortNameLower[$shortEndNameLower] = $this->resolveSymbol($symbol, $uses, $namespaceName);
+                    $symbol = substr($symbol, 0, strrpos($symbol, '\\'));
                 }
```

Comment:
answered in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7719#discussion_r1457787957 - I did my best to keep the test cases passing, the topic is otherwise quite complex, as when we shorten some FQCN, new conflicts might occur and some imports might be freed... `while` loop should be guaranteed to be left.